addons/mail/models/discuss/discuss_channel.py

```diff
@@ -1196,3 +1196,10 @@ def execute_command_who(self, **kwargs):
         else:
             msg = _("You are alone in this channel.")
         self.env.user._bus_send_transient_message(self, msg)
+
+    @api.model
+    def _get_mail_message_access(self, res_ids, operation, model_name=None):
+        if self.env.user._is_public():
+            if self.env["mail.guest"]._get_guest_from_context():
+                return "read"
+        return super()._get_mail_message_access(res_ids, operation, model_name)
```

Comment:
I'm still unsure about this part following our discussion. 

If we want to define it, conceptually it should always be `read`, not conditional to guest.
But we want more restrictions than that.

Reply:
As discussed, this part will remain to be managed in another PR.